File: workers/status/src/evidence/query.ts
/** 证据 locator query 的有限标量。/ Finite scalar admitted by evidence locator queries. */
export type LocatorQueryValue = string | number | boolean;

/** 顶层 TelemetryReference 提供的不可变查询身份。/ Immutable query identity supplied by the top-level TelemetryReference. */
export interface QueryIdentity {
  /** 权威 service name。/ Authoritative service name. */
  readonly serviceName: string;
  /** 权威 deployment UUIDv7。/ Authoritative deployment UUIDv7. */
  readonly deploymentId: string;
}

const LOG_KEYS = Object.freeze([
  "service",
  "environment",
  "deployment_id",
  "trace_id",
  "span_id",
  "severity",
  "operation",
  "component",
  "error_type",
] as const);
const METRIC_KEYS = Object.freeze([
  "service",
  "environment",
  "deployment_id",
  "region",
  "operation",
  "component",
  "dependency",
] as const);
const PROFILE_KEYS = Object.freeze([
  "service",
  "environment",
  "deployment_id",
  "region",
  "instance_id",
] as const);

const LABEL_NAMES: Readonly<Record<string, string>> = Object.freeze({
  service: "service_name",
  environment: "environment",
  deployment_id: "deployment_id",
  trace_id: "trace_id",
  span_id: "span_id",
  severity: "severity",
  operation: "operation",
  component: "component",
  error_type: "error_type",
  region: "region",
  dependency: "dependency",
  instance_id: "instance_id",
});

/** 将通用 SDK 类别映射到稳定的 Pyroscope 默认 sample type。/ Maps generic SDK categories to stable default Pyroscope sample types. */
const PROFILE_TYPE = Object.freeze({
  cpu: "process_cpu:cpu:nanoseconds:cpu:nanoseconds",
  memory: "memory:inuse_space:bytes:space:bytes",
  allocations: "memory:alloc_space:bytes:space:bytes",
  mutex: "mutex:delay:nanoseconds:contentions:count",
  goroutine: "goroutine:goroutine:count:goroutine:count",
  wall: "wall:wall:nanoseconds:cpu:nanoseconds",
} satisfies Readonly<Record<string, string>>);

const PROMETHEUS_METRIC = /^[a-zA-Z_:][a-zA-Z0-9_:]*$/u;
const PROMETHEUS_RESERVED_METRICS = new Set([
  "bool",
  "on",
  "ignoring",
  "group_left",
  "group_right",
]);
const MAX_COMPILED_LENGTH = 4_096;

/**
 * 将 SDK 固定键编译为带 service/deployment 范围的 LogQL stream selector。
 * Compiles SDK fixed keys into a LogQL stream selector scoped by service and deployment.
 *
 * 任意 `expression` 不被接受，因为无法在不解析完整 LogQL AST 时安全注入 identity scope。
 * Arbitrary `expression` is rejected because identity scope cannot be safely injected without a full LogQL AST parser.
 *
 * @see https://grafana.com/docs/loki/latest/query/query_reference/
 */
export function compileLogQuery(
  query: Readonly<Record<string, LocatorQueryValue>>,
  identity: QueryIdentity,
): string | undefined {
  return compileSelector(query, LOG_KEYS, identity);
}

/**
 * 将 SDK 固定键编译为 PromQL vector selector，不再静默丢弃 filters。
 * Compiles SDK fixed keys into a PromQL vector selector without silently dropping filters.
 *
 * @see https://prometheus.io/docs/prometheus/latest/querying/basics/
 */
export function compileMetricQuery(
  metricName: string,
  query: Readonly<Record<string, LocatorQueryValue>>,
  identity: QueryIdentity,
): string | undefined {
  if (
    !PROMETHEUS_METRIC.test(metricName) ||
    PROMETHEUS_RESERVED_METRICS.has(metricName)
  )
    return undefined;
  const selector = compileSelector(query, METRIC_KEYS, identity);
  if (selector === undefined) return undefined;
  return bounded(`${metricName}${selector}`);
}

/**
 * 将 SDK profile 类型与固定键编译为 Pyroscope profile selector。
 * Compiles an SDK profile type and fixed keys into a Pyroscope profile selector.
 *
 * 旧 render API 把 `profileId` 当任意 query 字符串，无法证明 identity scope，因此拒绝。
 * The legacy render API treats `profileId` as an arbitrary query string and cannot prove identity scope, so it is rejected.
 *
 * @see https://grafana.com/docs/pyroscope/latest/reference-server-api/
 */
export function compileProfileQuery(
  profileType: keyof typeof PROFILE_TYPE,
  profileId: string | undefined,
  query: Readonly<Record<string, LocatorQueryValue>> | undefined,
  identity: QueryIdentity,
): string | undefined {
  if (query === undefined || profileId !== undefined) return undefined;
  const selector = compileSelector(query, PROFILE_KEYS, identity);
  if (selector === undefined) return undefined;
  return bounded(`${PROFILE_TYPE[profileType]}${selector}`);
}

/** 编译确定顺序的相等 label matchers。/ Compiles deterministic equality label matchers. */
function compileSelector<const K extends string>(
  query: Readonly<Record<string, LocatorQueryValue>>,
  allowedKeys: readonly K[],
  identity: QueryIdentity,
): string | undefined {
  const keys = Object.keys(query);
  if (keys.length === 0 || keys.some((key) => !allowedKeys.includes(key as K)))
    return undefined;
  if (
    !matchesIdentity(query.service, identity.serviceName) ||
    !matchesIdentity(query.deployment_id, identity.deploymentId)
  )
    return undefined;

  const matchers = [
    `service_name=${quoteLabelValue(identity.serviceName)}`,
    `deployment_id=${quoteLabelValue(identity.deploymentId)}`,
  ];
  for (const key of allowedKeys) {
    if (key === "service" || key === "deployment_id") continue;
    const value = query[key];
    if (!isScalar(value)) return undefined;
    if (value !== undefined) {
      const labelName = LABEL_NAMES[key];
      if (labelName === undefined) return undefined;
      matchers.push(`${labelName}=${quoteLabelValue(String(value))}`);
    }
  }
  return bounded(`{${matchers.join(",")}}`);
}

/** 仅允许精确匹配权威 identity。/ Allows only an exact match with authoritative identity. */
function matchesIdentity(
  candidate: LocatorQueryValue | undefined,
  expected: string,
): boolean {
  return candidate === undefined || candidate === expected;
}

/** 使用 Go/LogQL/PromQL 双引号规则转义 label value。/ Escapes a label value using Go-compatible LogQL/PromQL double-quoted syntax. */
function quoteLabelValue(value: string): string {
  let output = '"';
  for (const character of value) {
    const code = character.charCodeAt(0);
    if (character === "\\") output += "\\\\";
    else if (character === '"') output += '\\"';
    else if (character === "\n") output += "\\n";
    else if (character === "\r") output += "\\r";
    else if (character === "\t") output += "\\t";
    else if (code < 0x20 || code === 0x7f)
      output += `\\x${code.toString(16).padStart(2, "0")}`;
    else if (code >= 0xd800 && code <= 0xdfff && character.length === 1)
      output += `\\u${code.toString(16).padStart(4, "0")}`;
    else output += character;
  }
  return `${output}"`;
}

/** 验证 locator 标量，undefined 只表示字段缺失。/ Validates locator scalars; undefined means only that a field is absent. */
function isScalar(
  value: LocatorQueryValue | undefined,
): value is LocatorQueryValue | undefined {
  return (
    value === undefined ||
    typeof value === "string" ||
    typeof value === "boolean" ||
    (typeof value === "number" && Number.isFinite(value))
  );
}

/** 应用编译查询的硬长度上限。/ Applies the hard compiled-query length bound. */
function bounded(value: string): string | undefined {
  return value.length <= MAX_COMPILED_LENGTH ? value : undefined;
}

File: crates/status-backend/src/evidence/query.rs
//! 身份限定的有限查询语言；不执行任意表达式。 / Identity-scoped finite query language; never executes arbitrary expressions.
use serde_json::Value;

/// 将固定键编译为部署限定选择器。 / Compile fixed keys into a deployment-scoped selector.
pub fn compile(reference: &Value) -> Option<String> {
    let kind = reference["kind"].as_str()?;
    let locator = &reference["locator"];
    let keys: &[&str] = match kind {
        "log_query" => &[
            "service",
            "environment",
            "deployment_id",
            "trace_id",
            "span_id",
            "severity",
            "operation",
            "component",
            "error_type",
        ],
        "metric_query" => &[
            "service",
            "environment",
            "deployment_id",
            "region",
            "operation",
            "component",
            "dependency",
        ],
        "profile" => &[
            "service",
            "environment",
            "deployment_id",
            "region",
            "instance_id",
        ],
        _ => return None,
    };
    let query = locator["query"].as_object()?;
    if query.is_empty() || query.keys().any(|k| !keys.contains(&k.as_str())) {
        return None;
    }
    let service = reference["service_name"].as_str()?;
    let deployment = reference["deployment_id"].as_str()?;
    for (key, expected) in [("service", service), ("deployment_id", deployment)] {
        if query.get(key).is_some_and(|v| v.as_str() != Some(expected)) {
            return None;
        }
    }
    let mut labels = vec![
        format!("service_name={}", quote(service)),
        format!("deployment_id={}", quote(deployment)),
    ];
    for key in keys
        .iter()
        .filter(|k| !["service", "deployment_id"].contains(k))
    {
        if let Some(v) = query.get(*key) {
            let text = match v {
                Value::String(s) => s.clone(),
                Value::Bool(_) | Value::Number(_) => v.to_string(),
                _ => return None,
            };
            labels.push(format!("{key}={}", quote(&text)));
        }
    }
    let prefix = match kind {
        "metric_query" => {
            let metric = locator["metric_name"].as_str()?;
            if metric.is_empty()
                || ["bool", "on", "ignoring", "group_left", "group_right"].contains(&metric)
                || !metric.bytes().enumerate().all(|(i, b)| {
                    b.is_ascii_alphabetic() || b"_:".contains(&b) || (i > 0 && b.is_ascii_digit())
                })
            {
                return None;
            }
            metric
        }
        "profile" => {
            if locator.get("profile_id").is_some() {
                return None;
            }
            match locator["profile_type"].as_str()? {
                "cpu" => "process_cpu:cpu:nanoseconds:cpu:nanoseconds",
                "memory" => "memory:inuse_space:bytes:space:bytes",
                "allocations" => "memory:alloc_space:bytes:space:bytes",
                "mutex" => "mutex:delay:nanoseconds:contentions:count",
                "goroutine" => "goroutine:goroutine:count:goroutine:count",
                "wall" => "wall:wall:nanoseconds:cpu:nanoseconds",
                _ => return None,
            }
        }
        _ => "",
    };
    let result = format!("{prefix}{{{}}}", labels.join(","));
    (result.encode_utf16().count() <= 4096).then_some(result)
}
/// Go 标签字符串转义。 / Go label-string escaping.
fn quote(s: &str) -> String {
    let mut out = String::from("\"");
    for c in s.chars() {
        match c {
            '\\' => out.push_str("\\\\"),
            '"' => out.push_str("\\\""),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if c < ' ' || c == '\u{7f}' => out.push_str(&format!("\\x{:02x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}
#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    #[test]
    fn identity_and_injection_boundaries() {
        let mut r = json!({"kind":"log_query","service_name":"api","deployment_id":"deploy","locator":{"query":{"severity":"error\"}"}}});
        assert_eq!(
            compile(&r).unwrap(),
            "{service_name=\"api\",deployment_id=\"deploy\",severity=\"error\\\"}\"}"
        );
        r["locator"]["query"]["expression"] = json!("{}");
        assert!(compile(&r).is_none());
        r["locator"]["query"] = json!({"service":"other"});
        assert!(compile(&r).is_none());
    }
}

#[cfg(test)]
mod security_tests {
    use super::*;
    use serde_json::json;
    #[test]
    fn accepts_only_finite_scoped_queries() {
        let mut r = json!({"kind":"metric_query","service_name":"api","deployment_id":"d","locator":{"metric_name":"requests_total","query":{"region":"eu","operation":"checkout"}}});
        assert_eq!(compile(&r).unwrap(),"requests_total{service_name=\"api\",deployment_id=\"d\",region=\"eu\",operation=\"checkout\"}");
        r["locator"]["metric_name"] = json!("requests_total or up");
        assert!(compile(&r).is_none());
        r["locator"]["metric_name"] = json!("on");
        assert!(compile(&r).is_none());
        r["kind"] = json!("profile");
        r["locator"] =
            json!({"profile_type":"cpu","profile_id":"arbitrary","query":{"service":"api"}});
        assert!(compile(&r).is_none());
        r["locator"].as_object_mut().unwrap().remove("profile_id");
        assert!(compile(&r)
            .unwrap()
            .starts_with("process_cpu:cpu:nanoseconds:cpu:nanoseconds{"));
        r["locator"]["query"] = json!({});
        assert!(compile(&r).is_none());
    }
    #[test]
    fn rejects_oversize_selectors() {
        let r = json!({"kind":"log_query","service_name":"api","deployment_id":"d","locator":{"query":{"operation":"x".repeat(4096)}}});
        assert!(compile(&r).is_none());
    }
}
